### PR TITLE
playwright-cli: align help with the runtime contract and close all official-CLI gaps

### PR DESCRIPTION
**Pull request by catalan-adobe**
## What

Two commits that make `playwright-cli`'s documented surface match what actually runs, then close every remaining gap against the official `@playwright/cli` schema.

### 1. Help ↔ runtime drift fixes (`cfd36c64`)

An agent following `--help` verbatim hit failures because the help documented forms the runtime rejects and omitted flags it requires:

- **`--tab` was missing from ~30 help entries** that require it (`resize <w> <h>` said nothing about `--tab`, then exited 1 asking for it). Every requiring entry now shows `--tab=<id>`, plus a global note under `Usage:`.
- **`screenshot --fullPage=true` was documented but rejected** — the known-flag walk only accepts boolean flags as the exact token. Now documented as `--fullPage|--full-page`.
- **`eval --output=<path>` validated cleanly but was never read**: exit 0, file never written. `eval` now honors `--output` as an alias of `--filename`, and `eval-file` honors `--filename`, so the sibling verbs agree.
- **8 registered verbs had no help entry** (`close`, `keydown`, `keyup`, `pdf`, `network-state-set`, `upload`, `state-save`, `state-load`) — their `<verb> --help` fell back to full help text that never mentioned them.
- New `help-drift.test.ts` pins the invariants: every manifest/handler verb has an extractable help entry, every manifest flag parses, every tab-requiring verb's entry shows `--tab`, and boolean flags are never documented with an attached value.

### 2. Official-CLI gap closure (`b8763d30`)

`playwright-cli-sync.mjs` reported 8 missing commands and 3 flag gaps. Implemented:

- `find [text]` / `find --regex=<re>` — search the page snapshot, returning matching lines with context (token-cheap alternative to a full snapshot)
- `snapshot --depth=<n>` — cap tree depth, with an explicit elision note
- `snapshot --boxes` — `[box=x,y,w,h]` viewport-relative rects on main-frame refs, resolved in one page evaluation
- `screenshot --type=png|jpeg|webp` — explicit or inferred from the `--filename` extension; unknown values rejected
- `screenshot --hires` — device-pixel capture via `clip.scale = devicePixelRatio`
- `open` / `tab-new --mobile` — generic mobile-device emulation (Pixel-class metrics + Android Chrome UA), sticky per target through the same re-apply-on-attach mechanism as `resize`

Skipped with reasons recorded in the sync tool: `tracing-*` (trace-viewer format is Playwright-internal; HAR recording covers network inspection), `video-*` (covered by SLICC's demo-recording tooling), `open --device` (`--mobile` covers the value without a named-device catalog). **The sync script now exits 0.**

## Bundle discipline

The feature bodies live in a lazily-imported `snapshot-features` chunk and the help text now loads on demand, so the kernel worker's eager first-load graph stays within the 4 kB per-change ratchet.

## Debt paydown

`state.ts` is off the layer-back-edge baseline (4 edges removed): `BrowserAPI`/`PageInfo`/`FrameInfo` are named via `playwright/types.ts`, the logger comes from `base/`, and the tray storage keys moved to `base/tray-storage-keys.ts` (re-exported from `scoops/` for existing callers). Baseline ratcheted 40→36 edges in 19→18 files.

## Docs & tests

- `SKILL.md` (agent-facing) and `help.ts` updated for every new command/flag; `docs/playwright-cli-sync.md` gap-state section refreshed
- `snapshot-features.test.ts` (16 tests) covers find/depth/boxes/type/hires/mobile including error paths; `validate-args.test.ts` updated for the now-implemented `--depth`/`--boxes`

## Verification

sync 0 gaps · `npm run verify` 7/7 · `check-touched-exemptions` clean · typecheck · 17,760 tests · coverage floors · both builds · bundle-size + first-load OK

### CODE REVIEW

**Comment by catalan-adobe** [comment]
### Strengths

- **`packages/webapp/tests/shell/supplemental-commands/playwright/help-drift.test.ts:38-86`** — Good structural regression guards now connect the manifest, registered handlers, parser vocabulary, required `--tab` flags, and boolean help syntax.
- **`packages/webapp/src/shell/supplemental-commands/playwright/handlers/snapshot.ts:128-165`** — `--depth` uses strict validation rather than permissive `parseInt`, and incompatible `--boxes --frame` usage fails explicitly.
- **`packages/webapp/src/shell/supplemental-commands/playwright-command.ts:166-205`** and **`packages/webapp/src/shell/supplemental-commands/playwright/handlers/snapshot.ts:29-37`** — Help and heavier snapshot features are genuinely lazy-loaded, keeping bundle concerns separated from feature behavior.
- **`packages/webapp/src/base/tray-storage-keys.ts:1-10`** and **`packages/webapp/src/shell/supplemental-commands/playwright/types.ts:14-20`** — The layer-back-edge cleanup is focused and preserves existing imports through re-exports without duplicating runtime values.
- **`packages/webapp/src/shell/supplemental-commands/playwright/handlers/snapshot.ts:281-286`** — Invalid screenshot formats fail before capture, rather than silently falling back to PNG.
- The sync script reports zero command, flag, argument, or undeclared-Slicc-only gaps for the reviewed head.

### Issues

#### 🚨 Critical (Must Fix Before Merge)

- **`packages/webapp/src/shell/supplemental-commands/playwright/handlers/tabs.ts:110-118`** — Mobile setup attaches to the new target outside the browser’s tab lock.
  - `BrowserAPI.setViewportOverride()` calls `attachToPage()` directly at `packages/webapp/src/cdp/browser-api.ts:285-300`, while `withTab()` is the mechanism that serializes access to the API’s mutable `sessionId` and transport (`browser-api.ts:230-251`).
  - Concrete failure: while another agent holds `withTab(tabB)` and is partway through a fill/click/eval, a concurrent `open --mobile` can attach `tabA` at line 111. The first command’s subsequent browser call can then use `tabA`’s session, potentially clicking or typing sensitive text into the wrong page.
  - Apply the identity and first navigation under one lock:
    ```ts
    await browser.withTab(targetId, async () => {
      await browser.setViewportOverride(targetId, width, height, mobileOptions);
      if (url !== 'about:blank') await browser.navigate(url);
    });
    ```
    Add a concurrency test using a controllable transport, rather than the current lock-free handler mock.

#### ⚠️ Important (Should Fix)

- **`packages/webapp/src/shell/supplemental-commands/playwright/handlers/snapshot-features.ts:50-64`** — `snapshot --boxes` cannot reliably associate refs with their elements.
  - The code discards every selector alternative after the first comma. More fundamentally, selectors are reconstructed from role/name. For a normal `<button>Buy</button>`, `buildRefSelector()` produces `button[aria-label="Buy"]` first (`packages/webapp/src/shell/supplemental-commands/playwright/snapshot.ts:32-41`), which does not match the text-labelled button. Duplicate accessible names also cause multiple refs to resolve to the same first element.
  - Consequently, common links/buttons omit boxes or receive another element’s coordinates even though the snapshot already carries reliable backend node IDs.
  - Pass and use `refToBackendNodeId`, or capture ref-to-rectangle identity as part of the snapshot traversal. If one page evaluation is mandatory, actual element identity must be established during the injected traversal rather than reconstructed afterward from role/name. Add tests with text-labelled and duplicate-name elements.

- **`packages/webapp/src/cdp/browser-api.ts:824-838`** — `--max-width` is incompatible with both new screenshot capabilities.
  - `pngWidth()` reads bytes 16–19 as a PNG IHDR without validating the signature (`browser-api.ts:28-40`). For JPEG/WebP output those bytes are unrelated data, so `--type=jpeg|webp --max-width=N` may stay oversized, attempt a nonsensical scale, or silently return the original after the caught recapture failure at lines 862-867.
  - For `--hires`, line 838 replaces the existing DPR scale. A 1280 CSS-pixel clip at DPR 2 initially produces 2560 pixels; limiting it to 1280 computes `0.5` and replaces scale `2`, yielding roughly 640 pixels rather than 1280. It should multiply the existing scale by the shrink ratio.
  - Use a format-aware image dimension reader and:
    ```ts
    existingClip.scale = (existingClip.scale ?? 1) * (maxWidth / encodedWidth);
    ```
    Add combination tests for JPEG/WebP plus `--max-width`, and `--hires --max-width` at DPR greater than one.

- **`packages/webapp/src/shell/supplemental-commands/playwright/handlers/tabs.ts:266-270`** — Resizing a mobile tab destroys its persisted mobile profile and leaves attachment-dependent behavior.
  - `setViewportOverride()` defaults omitted options to DPR 1 and `mobile: false`, then overwrites the saved override (`packages/webapp/src/cdp/browser-api.ts:292-300`). It does not clear the current session’s mobile UA, however.
  - Concrete failure: after `open --mobile`, running `resize` produces desktop metrics with the old mobile UA in the current session. After switching tabs and reattaching, the recorded override has no UA, so the same target becomes fully desktop. This contradicts the documented “sticky for the tab’s lifetime” behavior.
  - Store device identity separately from viewport dimensions, or provide an update-size operation that preserves the target’s existing DPR/mobile/UA settings. Test mobile → resize → attach another tab → reattach mobile.

#### 💡 Minor (Nice to Have)

- **`packages/webapp/src/shell/supplemental-commands/playwright/handlers/eval.ts:77-98`** — The two output aliases use opposite precedence when both are supplied.
  - `eval` chooses `--filename`, while `eval-file` chooses `--output`. Since validation permits both, otherwise equivalent invocations write to different paths.
  - Reject using both aliases together, or resolve them through one shared helper with consistent precedence.

- **`packages/webapp/src/shell/supplemental-commands/playwright/handlers/snapshot-features.ts:151-162`** — Overlapping `find` context windows are emitted repeatedly.
  - Accessible snapshots commonly have the same label on a control and its child static-text node. Adjacent matches therefore duplicate nearly identical five-line blocks, undermining the command’s token-saving purpose.
  - Merge overlapping context intervals before rendering and count either matched lines or rendered regions explicitly.

### Recommendations

- Add interaction-matrix tests rather than testing each new flag only in isolation: `type × max-width`, `hires × max-width`, mobile × resize/reattach, and boxes against real snapshot-generated selectors.
- Add a concurrency regression test that verifies every operation which calls `attachToPage()` participates in the tab lock.
- Consider making the official-schema sync check an automated dependency-update gate; currently future alignment still relies on a manual invocation.

### Assessment

**Ready to merge?** No

**Reasoning:** The overall structure, documentation work, lazy loading, and schema alignment are strong, but `open --mobile` currently bypasses the tab serialization contract and can redirect concurrent automation to the wrong page. Box resolution and screenshot flag composition also leave significant portions of the newly advertised behavior incorrect.

---
*Review generated by gh pr-review · model: github-copilot/gpt-5.6-sol*

File: packages/webapp/src/shell/supplemental-commands/playwright/handlers/tabs.ts [hunk 1]
```diff
@@ -7,21 +7,115 @@ import { fetchAndDiscover } from '../discover.js';
...
+  const teleStartStr = flags['teleport-start'];
+  const teleReturnStr = flags['teleport-return'];
+  if (!teleStartStr || !teleReturnStr) return null;
+
+  log.info('Arming teleport via open/tab-new flags');
+  log.debug('Arming teleport via open/tab-new flags details', {
+    targetId,
+    startPattern: teleStartStr,
+    returnPattern: teleReturnStr,
+  });
```
[the remark this answers is not in the document]
Comment (catalan-adobe) [reply packages/webapp/src/shell/supplemental-commands/playwright/handlers/tabs.ts:47]: Both addressed:

1. **UA-before-navigation** was fixed in 354f592d0 (this comment anchors at the fixed code): `--mobile` now creates the tab on `about:blank`, applies the emulation override, and only then navigates — the first document request carries the mobile UA. Test pins the ordering via `invocationCallOrder`.
2. **Touch emulation** added in fac1c29db: the mobile path of `applyViewportOverride` now sends `Emulation.setTouchEmulationEnabled { enabled: true, maxTouchPoints: 5 }`, re-applied on every fresh attach like the metrics/UA. Covered in `browser-api.test.ts`.

**Comment by catalan-adobe** [comment]
All review findings addressed in fac1c29db (each with a regression test):

**🚨 Critical — mobile setup outside the tab lock:** fixed exactly as suggested — the emulation override and first navigation now run under one `withTab(targetId)` hold. The test harness tracks which tab the lock is held for and asserts both `setViewportOverride` and `navigate` execute while it's held.

**⚠️ --boxes selector reconstruction:** rects now resolve through each ref's `backendNodeId` (`DOM.resolveNode` + `Runtime.callFunctionOn getBoundingClientRect`, no scrollIntoView) — the element identity the snapshot already recorded. Text-labelled elements and duplicate accessible names are covered by the updated test; a detached node leaves its line unannotated rather than lying.

**⚠️ --max-width composition:**
- `pngWidth` now validates the PNG signature (JPEG/WebP bytes no longer decode to a garbage width), and the handler rejects `--max-width` with non-PNG output loudly instead of silently returning the oversized original.
- With an existing clip (`--hires`), the downscale now composes: `scale = (existing ?? 1) × (maxWidth / encodedWidth)` — the 2560→1280 case yields scale 1, not 0.5. Both cases tested at the `BrowserAPI.screenshot` level with method-keyed mocks.

**⚠️ resize destroying the mobile profile:** `setViewportOverride` now inherits DPR/mobile/UA from the target's existing override when options are omitted — `resize` changes only the dimensions, and the merged override is what re-attach re-applies (tested: mobile → resize → tab-switch → reattach).

**💡 alias precedence:** both verbs now share `resolveOutputPath` and reject `--filename` + `--output` together.

**💡 find duplicate context:** overlapping/adjacent context windows are merged into regions before rendering; the header reports `N matching line(s) in M region(s)`.

Also from the inline thread: touch emulation (`Emulation.setTouchEmulationEnabled`, maxTouchPoints 5) is now part of the mobile path, re-applied on every fresh attach.

Not taken up in this PR: the sync-check-as-Renovate-gate recommendation — tracked in `docs/playwright-cli-sync.md` as the documented follow-up. Full local pass after the changes: verify 7/7, debt gate clean, typecheck, 17,767 tests, first-load OK.
